compiler/include/ForallStmt.h

```diff
@@ -31,6 +31,67 @@ enum ForallAutoLocalAccessCloneType {
   STATIC_AND_DYNAMIC
 };
 
+// Support for reporting calls that are not optimized for different reasons
+enum CallRejectReason {
+  CRR_ACCEPT,
+  CRR_NOT_ARRAY_ACCESS_LIKE,
+  CRR_NO_CLEAN_INDEX_MATCH,
+  CRR_ACCESS_BASE_IS_LOOP_INDEX,
+  CRR_ACCESS_BASE_IS_NOT_OUTER_VAR,
+  CRR_ACCESS_BASE_IS_COMPLEX_SHADOW_VAR,
+  CRR_ACCESS_BASE_IS_REDUCE_SHADOW_VAR,
+  CRR_TIGHTER_LOCALITY_DOMINATOR,
+  CRR_UNKNOWN,
+};
+
+class ALACandidate {
+  public:
+    ALACandidate() = delete;
+    ALACandidate(CallExpr *call, ForallStmt *forall, bool checkArgs=false);
+    inline CallExpr* getCall() const { return call_; }
+
+    inline int getIterandIdx() const { return iterandIdx_; }
+    inline void setIterandIdx(int iterandIdx) { iterandIdx_ = iterandIdx; }
+
+    inline CallRejectReason getReason() const { return reason_; }
+    inline void setReasonIfNeeded(CallRejectReason reason) {
+      if (!hasReason()) reason_ = reason;
```

Comment:
Should this check for CRR_ACCEPT? If it was previously accepted and a failure reason is passed in, it won't be set.

Reply:
Hmm, while I can't think of a way that's happening in current implementation, at a high-level that feels safe, and easy to add. I'll do it.

Reply:
https://github.com/chapel-lang/chapel/pull/25712/commits/a2485dd6d2236f3c209c6cf40a5e8e249cd40a32